File: Lowest Common Ancestor of a Binary Tree/solution2.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
var lowestCommonAncestor = function(root, p, q) {
  if (root === null) {
      return null
  }
  
  if (root.val === p.val || root.val === q.val) {
      return root
  }
  
  const leftNodeMatches = lowestCommonAncestor(root.left, p, q)
  const rightNodeMatches = lowestCommonAncestor(root.right, p, q)
  
  if (leftNodeMatches && rightNodeMatches) {
      return root
  }

  return leftNodeMatches || rightNodeMatches
};

File: 109. Convert Sorted List to Binary Search Tree/solution2.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

struct TreeNode *buildTree(struct ListNode **cur, int start, int end) {
    if (start > end) { return NULL; }
    int mid = (start + end) / 2;
    struct TreeNode *root = malloc(sizeof(struct TreeNode));
    root->left = buildTree(cur, start, mid - 1);
    root->val = (*cur)->val;
    *cur = (*cur)->next;
    root->right = buildTree(cur, mid + 1, end);
    return root;
}

struct TreeNode* sortedListToBST(struct ListNode* head) {
    int length = 0;
    struct ListNode *cur = head;
    while(cur) { ++length; cur = cur->next; }
    return buildTree(&head, 0, length - 1);
}